docs/architecture/engine.md

```diff
@@ -0,0 +1,270 @@
+   │     │    │  │                                  │   │     │      │
+   │     │    │  │                                  │   │     │      │
+   │     │    │  │                                  │   │     │      │
+   │     │    │  │          ┌───────────┐           │   │     │      │
+   │     │    │  └──────────┤ Series 1  ├───────────┘   │     │      │
+   │     │    │             └───────────┘               │     │      │
+   │     │    │                                         │     │      │
+   │     │    │                                         │     │      │
+   │     │    │                                         │     │      │
+   │     │    │                                         │     │      │
+   │     │    │                                         │     │      │
+   │     │    │                                         │     │      │
+   │     │    │                                         │     │      │
+   │     │    │                                         │     │      │
+   │     │    │                                         │     │      │
+   │     │    │               ┌───────┐                 │     │      │
+   │     │    └───────────────┤Shard 1├─────────────────┘     │      │
+   │     │                    └───────┘                       │      │
+   │     │             ┌──────────────────────┐               │      │
+   │     └─────────────┤     Namespace 1      ├───────────────┘      │
+   │                   └──────────────────────┘                      │
+   │                                                                 │
+   │                                                                 │
+   │                                                                 │
+   │              ┌───────────────────────────────┐                  │
+   └──────────────┤           Database            ├──────────────────┘
+                  └───────────────────────────────┘
+```
+
+The in-memory portion of M3DB is implemented via a hierarchy of datastructures:
+
+1. A "database" (one per M3DB node, effectively a singleton)
+
+2. "Namespaces" which are similar to tables or namespaces in other databases. A database "owns" numerous namespaces, and each namespace has a unique name as well as distinct configuration with regards to data retention and blocksize (which we will discuss in more detail later).
+
+3. ["Shards"](sharding.md) which are owned by namespaces. Shards are effectively the same as "virtual shards" in Cassandra in that they provide arbitrary distribution of time series data via a simple hash of the series ID. Shards are useful through the entire M3DB stack in that they make horizontal scaling and adding / removing nodes without downtime trivial at the cluster level, as well as providing more fine grained lock granularity at the memory level, and finally they inform the filesystem organization in that data belonging to the same shard will be used / dropped together and can be kept in the same file.
```

Comment:
Lots of good detail here, but probably better to break out some of the nuance into the sharding.md file?

Reply:
good call